drltrace/drltrace.c

```diff
@@ -0,0 +1,369 @@
+    const char *modname = NULL;
+    app_pc func = drwrap_get_func(wrapcxt);
+    module_data_t *mod;
+    if (options.only_from_app) {
+        /* For just this option, the modxfer approach might be better */
+        app_pc retaddr =  NULL;
+        void *drcontext = drwrap_get_drcontext(wrapcxt);
+        DR_TRY_EXCEPT(drcontext, {
+            retaddr = drwrap_get_retaddr(wrapcxt);
+        }, { /* EXCEPT */
+            retaddr = NULL;
+        });
+        if (retaddr != NULL) {
+            mod = dr_lookup_module(retaddr);
+            if (mod != NULL) {
+                bool from_exe = (mod->start == exe_start);
+                dr_free_module_data(mod);
+                if (!from_exe)
+                    return;
+            }
+        } else {
+            /* Nearly all of these cases should be things like KiUserCallbackDispatcher
+             * or other abnormal transitions.
+             * If the user really wants to see everything they can not pass
+             * -only_from_app.
+             */
+            return;
+        }
+    }
+    /* XXX: it may be better to heap-allocate the "module!func" string and
+     * pass in, to avoid this lookup.
+     */
+    mod = dr_lookup_module(func);
+    if (mod != NULL)
+        modname = dr_module_preferred_name(mod);
+    dr_fprintf(outf, "%s%s%s%s", (outf == STDERR ? STDERR_PREFIX : ""),
+               modname == NULL ? "" : modname,
+               modname == NULL ? "" : "!", name);
+    if (options.all_args > 0) {
+        uint i;
+        void *drcontext = drwrap_get_drcontext(wrapcxt);
+        dr_fprintf(outf, "(");
+        DR_TRY_EXCEPT(drcontext, {
+            for (i = 0; i < options.all_args; i++) {
+                dr_fprintf(outf, "%s"PFX, (i != 0) ? ", " : "",
+                           drwrap_get_arg(wrapcxt, i));
+            }
+        }, {
+            dr_fprintf(outf, "<invalid memory>");
+            /* Just keep going */
+        });
+        dr_fprintf(outf, ")");
+    }
+    dr_fprintf(outf, "\n");
+    if (mod != NULL)
+        dr_free_module_data(mod);
+}
+
+static void
+iterate_exports(const module_data_t *info, bool add)
+{
+    dr_symbol_export_iterator_t *exp_iter =
+        dr_symbol_export_iterator_start(info->handle);
+    while (dr_symbol_export_iterator_hasnext(exp_iter)) {
+        dr_symbol_export_t *sym = dr_symbol_export_iterator_next(exp_iter);
+        app_pc func = NULL;
+        if (sym->is_code)
+            func = sym->addr;
+#ifdef LINUX
+        else if (sym->is_indirect_code) {
+            /* Invoke the export to get the real entry: */
+            app_pc (*indir)(void) = (app_pc (*)(void)) cast_to_func(sym->addr);
+            void *drcontext = dr_get_current_drcontext();
+            DR_TRY_EXCEPT(drcontext, {
+                func = (*indir)();
+            }, { /* EXCEPT */
+                func = NULL;
+            });
+            NOTIFY(1, "export %s indirected from "PFX" to "PFX""NL,
```

Comment:
OK, you mean runsuite.cmake's checks, makes sense (maybe say that in the commit msg).